server/statistics/topn_test.go

```diff
@@ -25,118 +26,238 @@ var _ = Suite(&testTopNSuite{})
 type testTopNSuite struct{}
 
 type item struct {
-	id    uint64
-	value float64
+	id     uint64
+	values []float64
 }
 
 func (it *item) ID() uint64 {
 	return it.id
 }
 
-func (it *item) Less(than TopNItem) bool {
-	return it.value < than.(*item).value
+func (it *item) Less(k int, than TopNItem) bool {
+	return it.values[k] < than.(*item).values[k]
 }
 
 func (s *testTopNSuite) TestPut(c *C) {
 	const Total = 10000
+	const K = 3
 	const N = 50
-	tn := NewTopN(N, 1*time.Hour)
-	for _, x := range rand.Perm(Total) {
-		c.Assert(tn.Put(&item{id: uint64(x), value: float64(-x) + 1}), IsFalse)
-	}
-	for _, x := range rand.Perm(Total) {
-		c.Assert(tn.Put(&item{id: uint64(x), value: float64(-x)}), IsTrue)
-	}
-	c.Assert(tn.GetTopNMin().(*item), DeepEquals, &item{id: N - 1, value: 1 - N})
-	topns := make([]float64, N)
-	for _, it := range tn.GetAllTopN() {
-		it := it.(*item)
-		topns[it.id] = it.value
+	tn := NewTopN(K, N, 1*time.Hour)
+
+	putPerm(c, tn, K, Total, func(x int) float64 {
+		return float64(-x) + 1
+	}, false /*insert*/)
+
+	putPerm(c, tn, K, Total, func(x int) float64 {
+		return float64(-x)
+	}, true /*update*/)
+
+	// check GetTopNMin
+	for k := 0; k < K; k++ {
+		c.Assert(tn.GetTopNMin(k).(*item).values[k], Equals, float64(1-N))
 	}
-	for i, v := range topns {
-		c.Assert(v, Equals, float64(-i))
+
+	{
+		topns := make([]float64, N)
+		// check GetAllTopN
+		for _, it := range tn.GetAllTopN(0) {
+			it := it.(*item)
+			topns[it.id] = it.values[0]
+		}
+		// check update worked
+		for i, v := range topns {
+			c.Assert(v, Equals, float64(-i))
+		}
 	}
-	all := make([]float64, Total)
-	for _, it := range tn.GetAll() {
-		it := it.(*item)
-		all[it.id] = it.value
+
+	{
+		all := make([]float64, Total)
+		// check GetAll
+		for _, it := range tn.GetAll() {
+			it := it.(*item)
+			all[it.id] = it.values[0]
+		}
+		// check update worked
+		for i, v := range all {
+			c.Assert(v, Equals, float64(-i))
+		}
 	}
-	for i, v := range all {
-		c.Assert(v, Equals, float64(-i))
+
+	{ // check all dimensions
+		for k := 1; k < K; k++ {
+			topn := make([]float64, 0, N)
+			for _, it := range tn.GetAllTopN(k) {
+				topn = append(topn, it.(*item).values[k])
+			}
+			sort.Sort(sort.Reverse(sort.Float64Slice(topn)))
+
+			all := make([]float64, 0, Total)
+			for _, it := range tn.GetAll() {
+				all = append(all, it.(*item).values[k])
+			}
+			sort.Sort(sort.Reverse(sort.Float64Slice(all)))
+
+			c.Assert(topn, DeepEquals, all[:N])
+		}
 	}
+
+	// check Get
 	for i := uint64(0); i < Total; i++ {
 		it := tn.Get(i).(*item)
 		c.Assert(it.id, Equals, i)
-		c.Assert(it.value, Equals, -float64(i))
+		c.Assert(it.values[0], Equals, -float64(i))
+	}
+}
+
+func putPerm(c *C, tn *TopN, K, Total int, f func(x int) float64, isUpdate bool) {
```

Comment:
`K, Total` should start with the lowercase.

Reply:
Renamed.